proposals/binary-compat-only.md

```diff
@@ -0,0 +1,123 @@
+# BinaryCompatOnlyAttribute
+
+## Summary
+[summary]: #summary
+
+We introduce a new attribute, `System.BinaryCompatOnlyAttribute`, that causes the type or member to which it is applied to be entirely inaccessible from source code.
+It will still be emitted as declared, with the accessibility declared, but will be treated as if it does not exist by all other compile-time C# mechanisms.
+
+## Motivation
+[motivation]: #motivation
+
+API authors often run into an issue of what to do with a member after it has been obsoleted. For backwards compatibility purposes, many will keep the existing member around
+with `ObsoleteAttribute` set to error in perpetuity, in order to avoid breaking consumers who upgrade binaries at runtime. This particularly hits plugin systems, where the
+author of a plugin does not control the environment in which the plugin runs. The creator of the environment may want to keep an older method present, but block access to it
+for any newly developed code. However, `ObsoleteAttribute` by itself is not enough. The type or member is still visible in overload resolution, and may cause unwanted overload
+resolution failures when there is a perfectly good alternative, but that alternative is either ambiguous with the obsoleted member, or the presence of the obsoleted member causes
+overload resolution to end early without ever considering the good member. For this purpose, we want to have a way to mark such members as technically being present in the DLL,
+but not visible to any code, so that they will have no impact on member lookup, overload resolution, or any other similar elements.
+
+## Detailed design
+[design]: #detailed-design
+
+### `System.BinaryCompatOnlyAttribute`
+
+We introduce a new reserved attribute:
+
+```cs
+namespace System;
+
+// Excludes Assembly, GenericParameter, Module, Parameter, ReturnValue
+[AttributeUsage(AttributeTargets.Class
+                | AttributeTargets.Constructor
+                | AttributeTargets.Delegate
+                | AttributeTargets.Enum
+                | AttributeTargets.Event
+                | AttributeTargets.Field
+                | AttributeTargets.Interface
+                | AttributeTargets.Method
+                | AttributeTargets.Property
+                | AttributeTargets.Struct,
+                AllowMultiple = false,
+                Inherited = false)]
+public class BinaryCompatOnlyAttribute : Attribute {}
+```
+
+When applied to a type member, that member is treated as inaccessible in every location by the compiler, meaning that it does not contribute to member
```

Comment:
Can a BinaryCompatOnly member reference another BinaryCompatOnly member?
e.g,

```csharp
[BinaryCompatOnly]
public class C1 { }

[BinaryCompatOnly]
public class C2 : C1 { }
```

Reply:
I think that should be possible, just like it is with `[Obsolete]` elements referring to other `[Obsolete]` elements

Reply:
As currently proposed, no. This is one of several open questions that I into.